python/src/opendp/extras/polars/__init__.py

```diff
@@ -848,6 +853,70 @@ def with_keys(
             self._query,
         )
 
+    def truncate_per_group(
+        self,
+        k: int,
+        by: list[Any] | None = None,
+        keep: Literal["first", "last", "sample"] = "first",
+    ) -> LazyFrameQuery:
+        """
+        Limit the number of contributed rows per group.
+
+        :param k: the number of rows to keep for each identifier and group
+        :param by: optional, additional columns to group by
+        :param keep: which rows to keep for each identifier in each group
+        """
+        input_metric = self._query._chain[1]
+
+        if isinstance(input_metric, FrameDistance):
+            input_metric = input_metric.inner_metric
+        if not isinstance(input_metric, (SymmetricIdDistance, ChangeOneIdDistance)):
+            raise ValueError("truncation is only valid when identifier is defined")
+
+        if keep == "first":
+            indexes = pl.int_range(pl.len())
+        elif keep == "last":
+            indexes = pl.int_range(pl.len()).reverse()
+        elif keep == "sample":
+            indexes = pl.int_range(pl.len()).shuffle()
```

Comment:
Given the attention we've given to the PRNG in other places, are we happy using what polars provides here?

Reply:
In this case, even if it were completely broken, it would still be fine.

Reply:
Last thought here: if you're concerned that shuffle will be slow, maybe the algorithm could change? if the group is small, we can just take the whole thing, and if the group is large, possibly quicker to generate just k distinct indices rather than a permutation of the whole group?

But this could be premature optimization!

Reply:
Switched to shuffle for both group-wise contributions and group contributions.